README.md
# Proteoglycan pathway evolution analysis

### Description
A collection of files associated with a project I contributed to as as final-year Bioinformatics student at the 
University of Saskatchewan.    

### Abstract
Proteoglycans are a class of macromolecules comprised of proteins to which 
carbohydrate groups are attached. They serve important functions in the 
synthesis of bone and cartilage. In addition, proteoglycans are known to 
function as extracellular signal molecules and have been identified 
intracellularly as well. The loss of proteoglycans is associated with 
debilitating diseases such as osteoarthritis. Therefore, an enriched knowledge
 of proteoglycans may aid osteoarthritis research by suggesting significant 
genes of the synthesis pathway, which can become subjects for more detailed 
studies. This research focused on collecting *Homo sapiens* orthologs 
of the proteoglycan synthesis pathway genes. Data for this study were 
collected from Ensembl. Multiple sequence alignments were performed to 
investigate gene sequence conservation across organisms. Phylogenetic analyses
 and an investigation of non-synonymous vs. synonymous mutation ratios of 
proteoglycan synthesis genes were conducted with the intent of verifying 
whether organisms that synthesize bone and cartilage exhibit higher mutation 
ratios compared to organisms that do not synthesize bone and cartilage. 
Organisms that synthesize bone and cartilage were hypothesized to showcase 
higher non-synonymous vs. synonymous mutation ratios compared to organisms
 that do not as a consequence of evolutionary pressures caused by
 environmental factors, such as the transition from water to land. Contrary 
to expectation, the research identified that organisms that synthesize bone 
and cartilage exhibit lower non-synonymous vs. synonymous mutation ratios 
compared to organism that do not synthesize bone and cartilage.

#### Presentations
See first presentation [here](https://github.com/flaviuvadan/proteoglycan-pathway-evolution/blob/master/presentations/proteglycan_pathway_conservation_part_1.pdf).


CHANGELOG.md
# Documentation of project progress

## March 19th, 2020
- forgot to update the log so I am doing this today
- worked on writing the paper and uploaded to git

## January 11th, 2020
- created a visualization of the box plots of each gene per habitat organism collection
- isolated organism classification namespaces into their own file 
- implemented the visualization of the box plots for organisms binned by bone classification

## January 9th, 2020
- found a way to showcase differences between grouped organisms
- started implementing the visualizing function of the dnds/visualizer

## January 8th, 2020
- started building the distributor package grouped visualizer method 
- still have to figure out a nice way to aggregate data so that it's nice to create subplots of it

## January 7th, 2020
- added more significant organisms based on feedback
- changed the logic of gaps vs. illegal characters in dnds/visualizer
- ran visualizer for all the significant organisms again
- modified dnds/loader to account for inf values (possible due to divergence)
- changed dnds/distributor to create subplots for all new orgs

## Dec 29th, 2019
- implemented the rest of the dnds visualizer
- fixed the dnds package (inconsistent naming of organisms between alignments and dnds was causing dnds scores to look 
odd)
- implemented a distribution visualization for the dN/dS values of all the significant organisms

## Dec 28th, 2019
- implemented the similarity mapper (uses Jaccard's index for similarity)
- implemented tree visualization for tree similarities 
- reimplemented the dnds package, turns out the dnds library does not work properly, will have to use the 
`ape` package in R (has 35 citations), turn the dnds package into a parsing and visualizing package, and
construct a pipe that calls the parsing script, calls R on the resulting files, then visualizes the dnds results
computed by the R `ape` package

## Dec 27th, 2019
- add more taxonomic details to the significant organisms phylogenetic tree
- performed a comparison against BLAST of organisms that are reported to not have a gene
- implemented a parser for the JSON files of the BLAST comparison

## Dec 26th, 2019
- chose and added a subset of the organisms that represent the classes of bones/cartilage for the study
- added the phylogenetic mapper for significant organisms

## Dec 3, 2019
- added first presentation

## Nov 28, 2019
- fixed an inaccurate piece of information regarding gene functions
- adjusted gene function chart to have a tight bounding box
- adjusted the gene frequencies chart - sorted organisms based on gene frequency 

## Nov 18, 2019
- start building the dnds package

## Nov 17, 2019
- adjusted MSA visualization to start x-axis at 0
- adjusted grouping of organisms based on taxonomic information

## Nov 4, 2019
- fixed the problem of false negative matches when building phylogeny trees for all the genes

## Nov 3, 2019
- added a phylogeny mapper for building the trees
- reorganized some data files
- wrote script for phylogenetic tree visualizations

## Nov 2, 2019
- implemented a taxon mapper for generating a plot of frequency of organisms per clade
- implemented a gene function visualizer

## Nov 1, 2019
- implemented the MSA parser
- moved some results files around (plots moved to src/data)

## Oct 31, 2019
- started building the MSAs parser
- noticed the alignments had a problem caused by the ClustalW format (limited number of characters for the organism 
name). Performed MSAs again, with Kalign on EBI, and saved in FASTA format 

## Oct 24, 2019
- make taxa collector build organism information files (gene frequency and genes per organism)

## Oct 13, 2019
- used EBI's Kalign to perform multiple sequence alignments for all genes
- added a test parser script for writing MSAs as single lines (will not use yet as they are challenging to interpret if 
written on single lines)
- built the taxa collector and collected organisms' information

## Oct 3, 2019
- moved exceptions out of orthologs package
- create curator package to parse homology information

## Oct 2, 2019
- curated gene Ensembl IDs list
- added orthologs package
- implemented Collector for calls to Ensembl

## Sep 21, 2019
- took the gene list form Brian and stored it in `genes.txt`
- used a pipe to clean up the list from the format:

```
gene1, gene2, ...
```
to

```
gene1 

gene2
```
via 

```
echo "gene1, gene2" | tr ',\s' '\n'
``` 
- looked into setting up a BLAST server and found [this resource](https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=Download)
- the objective is to find how a given set of genes (G) relate to cartilage production in multiple organisms (O)
- O will contain:

1. Organisms that make bone
1. Organisms that only produce cartilage
1. Produce neither cartilage nor bone
1. Used to have bone but now make cartilage (note, this is different than *only* producing cartilage)

- see `genes.txt` for G
- explored the [Ensembl Species Tree](https://uswest.ensembl.org/info/about/speciestree.html) to identify what organisms to use in the project (see `organisms.txt`)
- found the [REST documentation of Ensembl](https://rest.ensembl.org/documentation/info/homology_ensemblgene)
- since the objective at the start of the project is to find genomes of species that host the genes of interest, all the project has to do at the start is find the Ensembl IDs of the genes of interest and use `GET homology/id/:id` to get orthologs (the response includes organisms, which can then be grouped into bone/cartilage/neither/used to have bone)

#### Potential programs to write
- parse the list of genes to get orthologs of each one, including species (using an Ensembl API);
- use the species information from 1 to hit some other API (taxonomy API from NCBI?) to identify vertebrates, chondrichthyes, etc and group them accordingly;

src/dnds/distributor.py
import os

import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

from src import exceptions
from src.dnds import organisms_classes


class Distributor:
    """ Responsible for reading in the dnds data to create distributions """

    ORG_CLASS_IDX = 0
    ORG_NAME_IDX = 1

    # number of genes in this study
    NUM_GENES = 51

    def __init__(self):
        """ Constructor """
        self.sig_orgs = self._get_significant_orgs()
        self.dnds_data = self._get_hg_data()
        self.habitat_bins = self._get_binned_habitat_dataframe(class_type=organisms_classes.OrganismHabitatGroups())
        self.bone_bins = self._get_binned_habitat_dataframe(class_type=organisms_classes.OrganismBoneGroups())

    def _get_significant_orgs(self):
        """ Reads in the significant organisms selected for this study """
        orgs = {}
        path = os.path.join(os.getcwd(), "src", "data", "phylogeny", "significant_organisms.txt")
        with open(path, "r") as f:
            f.readline()  # don't care about the top line
            for line in f.readlines():
                org_name = line.split(",")[self.ORG_NAME_IDX]
                org = "_".join(org_name.lower().split())
                orgs[org] = 1
        return orgs

    def _get_hg_data(self):
        """ Creates a dictionary of the data to be plotted """
        final = {}
        for sig_org in self.sig_orgs.keys():
            sig_org_csv = os.path.join(os.getcwd(), "src", "data", "dnds", "{}.csv".format(sig_org))
            df = pd.read_csv(sig_org_csv)
            histo_vals = []
            for x in df.values:
                histo_vals = histo_vals + list(x[1:])
            final[sig_org] = sorted(histo_vals)
        return final

    def visualize(self):
        """ Creates a histogram of the independent significant organisms dN/dS distribution along with a histogram of
        the binned version for visualizing the distribution per group """
        self._visualize_independent_orgs()
        self._visualize_by_habitat()
        self._visualize_by_bone_class()

    def _visualize_independent_orgs(self):
        """ Creates a collection of distributions, one for each significant organism. The distributions represent the
        pattern in dnds ratios for each organism """
        fig, ax = plt.subplots(7, 3,
                               figsize=(12, 18))
        # fig.suptitle("Significant organisms' dN/dS distributions",
        #              fontsize=15)
        for idx, k in enumerate(self.dnds_data.keys()):
            plt.subplot(7, 3, idx + 1)
            fonts = {
                'fontsize': 11,
                'fontweight': 2,
                'verticalalignment': 'baseline',
                'horizontalalignment': 'center'
            }
            plt.title(" ".join(k.split("_")[:2]).capitalize(),
                      fontdict=fonts)
            data = self.dnds_data.get(k)
            lt_1 = len([x for x in data if x < 1])
            gt_1 = len(data) - lt_1
            plt.axis([0, 7, 0, 800])
            # have to remove indices manually since I cannot figure out why sharex and sharey have no effect when used
            # with plt.subplots
            no_x_tick = [x for x in range(1, 19)]
            if idx + 1 in no_x_tick:
                plt.xticks([])
            y_tick = [y for y in range(1, 20, 3)]
            if idx + 1 not in y_tick:
                plt.yticks([])
            plt.hist(self.dnds_data.get(k), bins=21)
            plt.axvline(1,
                        color='k',
                        linestyle='dashed',
                        linewidth=1)
            # I hate arbitrary numbers like this, Starman
            plt.annotate("{} < 1\n{} >= 1".format(lt_1, gt_1), (5, 550))
        title = os.path.join(os.getcwd(), "src", "data", "visualizations", "dnds", "grouped_orgs", "histograms.pdf")
        plt.savefig(title, format="pdf", quality=95, bbox_inches='tight')

    def _get_binned_habitat_dataframe(self, class_type=None):
        """ Constructs and returns the cumulative dataframes of the significant organisms based on the given class
         namespace organization """
        if not class_type:
            raise exceptions.EmptyNamespaceClassException('namespace class not given')
        binned = {}
        for org_class, orgs in class_type.class_org_map.items():
            binned[org_class] = {}
            main_df = pd.DataFrame()
            for org in orgs:
                org_csv = os.path.join(os.getcwd(), "src", "data", "dnds", "{}.csv".format(org))
                df = pd.read_csv(org_csv)
                main_df = pd.concat([main_df, df])
            binned[org_class] = main_df
        return binned

    def _visualize_by_habitat(self):
        """ Creates a collection of distributions for each organism habitat. The bins are: terrestrial, aquatic,
        terrestrial and aquatic """
        groups = organisms_classes.OrganismHabitatGroups()
        fig, (ax1, ax2, ax3) = plt.subplots(1, 3, figsize=(15, 15))
        ax1.set_title('Terrestrial organisms')
        ax2.set_title('Aquatic organisms')
        ax3.set_title('Terrestrial/aquatic organisms')

        ax1.set_xlim(0, 6)
        ax2.set_xlim(0, 6)
        ax3.set_xlim(0, 6)

        terr_df = self.habitat_bins.get(groups.TERR)
        terr_df = terr_df.drop(columns='organism')
        aqua_df = self.habitat_bins.get(groups.AQUA)
        aqua_df = aqua_df.drop(columns='organism')
        traq_df = self.habitat_bins.get(groups.TERR_AQUA)
        traq_df = traq_df.drop(columns='organism')

        sns.violinplot(data=terr_df, orient='h', color='deepskyblue', ax=ax1)
        sns.violinplot(data=aqua_df, orient='h', color='deepskyblue', ax=ax2)
        sns.violinplot(data=traq_df, orient='h', color='deepskyblue', ax=ax3)

        plt.subplots_adjust(wspace=0.3)
        save_path = os.path.join(os.getcwd(), 'src', 'data', 'visualizations', 'dnds', 'grouped_orgs',
                                 'habitat_dist.pdf')
        plt.savefig(save_path, dpi=95, bbox_inches='tight')

    def _visualize_by_bone_class(self):
        """ Creates a collection of distributions for each organism bin. The bins are: bone and cartilage, cartilage
        only, neither """
        groups = organisms_classes.OrganismBoneGroups()
        fig, (ax1, ax2, ax3) = plt.subplots(1, 3, figsize=(15, 15))
        ax1.set_title('Bony organisms')
        ax2.set_title('Cartilaginous organisms')
        ax3.set_title('Organisms without bone or cartilage')

        ax1.set_xlim(0, 6)
        ax2.set_xlim(0, 6)
        ax3.set_xlim(0, 6)

        terr_df = self.bone_bins.get(groups.BONE_CART)
        terr_df = terr_df.drop(columns='organism')
        aqua_df = self.bone_bins.get(groups.CART_ONLY)
        aqua_df = aqua_df.drop(columns='organism')
        traq_df = self.bone_bins.get(groups.NO_BONE_NO_CART)
        traq_df = traq_df.drop(columns='organism')

        sns.violinplot(data=terr_df, orient='h', color='deepskyblue', ax=ax1)
        sns.violinplot(data=aqua_df, orient='h', color='deepskyblue', ax=ax2)
        sns.violinplot(data=traq_df, orient='h', color='deepskyblue', ax=ax3)

        plt.subplots_adjust(wspace=0.3)
        save_path = os.path.join(os.getcwd(), 'src', 'data', 'visualizations', 'dnds', 'grouped_orgs', 'bone_dist.pdf')
        plt.savefig(save_path, dpi=95, bbox_inches='tight')


if __name__ == "__main__":
    dist = Distributor()
    dist.visualize()
